pkg/services/x509/x509_validator.go

```diff
@@ -0,0 +1,270 @@
+	if err = json.Unmarshal(headerStr, headers); err != nil {
+		return nil, fmt.Errorf("could not unmarshall headers: %w", err)
+	}
+
+	certsFromHeader := headers.X509CertChain()
+	if len(certsFromHeader) == 0 {
+		return nil, fmt.Errorf("the jwt x5c field should contain at least 1 certificate")
+	}
+
+	chain, err := validator.parseCertsFromHeader(certsFromHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	certUsedForSigning := chain[0]
+
+	payload, err := jws.Verify([]byte(rawAuthToken), headers.Algorithm(), certUsedForSigning.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify jwt signature: %w", err)
+	}
+
+	parsedJwt := jwt.New()
+	if err := json.Unmarshal(payload, &parsedJwt); err != nil {
+		return nil, fmt.Errorf("could not parse jwt payload: %w", err)
+	}
+
+	return &JwtX509Token{
+		chain:  chain,
+		token:  parsedJwt,
+		raw:    rawAuthToken,
+		sigAlg: headers.Algorithm(),
+	}, nil
+}
+
+func (validator JwtX509Validator) parseCertsFromHeader(certsFromHeader []string) ([]*x509.Certificate, error) {
+	// Parse all the certificates from the header into a chain
+	chain := []*x509.Certificate{}
+
+	for _, certStr := range certsFromHeader {
+		rawCert, err := base64.StdEncoding.Strict().DecodeString(certStr)
+		if err != nil {
+			return nil, fmt.Errorf("could not base64 decode cert: %w", err)
+		}
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse x509 certificate: %w", err)
+		}
+		chain = append(chain, cert)
+	}
+	return chain, nil
+}
+
+func (validator JwtX509Validator) Verify(x509Token *JwtX509Token) error {
+	var sigAlgAllowed bool
+	for _, allowedAlg := range validator.allowedSigningAlgs {
+		if allowedAlg == x509Token.sigAlg {
+			sigAlgAllowed = true
+			break
+		}
+	}
+	if !sigAlgAllowed {
+		return fmt.Errorf("signature algorithm %s is not allowed", x509Token.sigAlg)
+	}
+
+	leafCert, verifiedChains, err := validator.verifyCertChain(x509Token)
+	if err != nil {
+		return err
+	}
+
+	for _, verifiedChain := range verifiedChains {
+		err := validator.checkCertRevocation(verifiedChain)
+		if err != nil {
+			return err
+		}
+	}
+
+	// parse the jwt and verify the jwt signature
+	token, err := jwt.ParseString(x509Token.raw, jwt.WithVerify(x509Token.sigAlg, leafCert.PublicKey))
+	if err != nil {
+		return err
+	}
+
+	// verify the jwt contents such as expiration
+	if err := jwt.Verify(token); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (validator JwtX509Validator) verifyCertChain(x509Token *JwtX509Token) (*x509.Certificate, [][]*x509.Certificate, error) {
+	rootsPool := x509.NewCertPool()
+	for _, cert := range validator.roots {
+		rootsPool.AddCert(cert)
+	}
+	intermediatesPool := x509.NewCertPool()
+	for _, cert := range validator.intermediates {
+		intermediatesPool.AddCert(cert)
+	}
+	verifyOpts := x509.VerifyOptions{
+		Intermediates: intermediatesPool,
+		Roots:         rootsPool,
+		CurrentTime:   time.Now(),
```

Comment:
doesn't the token have issued/signing time? Since certificate validity should be valid at the moment of signing, not the moment of checking. Also, I think Golang X.509 API is a bit vague on this point since they're calling it 'CurrentTime'. Probably because when it was written it was aimed on TLS connections (and not signing).

Reply:
Interesting point. Should you be able to sign a login token with a validity beyond the certificate validity? This check is meant to validate if the token is currently valid. But perhaps we should use the issued at (iat) value from the jwt?

Reply:
I think we should indeed use `iat`. Certificates shouldn't be used for new signatures after their validity has expired.